design/authenticators/application-identity/application-identity.md

```diff
@@ -0,0 +1,464 @@
+       granularity than what is expected by the user.
+  - Validate required constraints exist
+    - For example, in the Azure Authenticator we require hosts to have the
+     `subscription-id` and the `resource-group` annotations. In the K8s
+      Authenticator we require the `namespace` annotation.
+  - Validate constraint combinations
+    - For example, in the Azure Authenticator we do not allow hosts to have
+     both the `user-assigned-identity` and the `system-assigned-identity`
+     annotations. In the K8s we do not allow hosts to have any combination of
+      the annotations `deployment`, `deployment-config` & `stateful-set`.
+      
+Also here we can see some generalization. 
+In this case the general class will be defined as follows:
+- input
+  - Different lists of annotation names 
+    - permitted
+    - required
+    - non-permitted combinations
+  - The `constraints` hash built in the previous step
+- output
+  - Nothing if validation succeeded, relevant error if it failed
+  
+#### Validate the Application Identity matches the authentication request
+
+Once we extracted the Application Identity from the host's annotations, and
+verified it is defined correctly, we can validate that the authentication
+request is valid by validating the Application Identity. For example, if
+the k8s host is permitted to authenticate from namespace `namespace-1`, and
+ the authentication request was sent from a pod that is in namespace
+`namespace-2` we will fail the request.
+
+In a way, we can define this step as comparing 2 objects of Application
+Identity - one that was extracted from the host and one that was extracted
+from the authentication request. In the example above we will have the
+ following Application Identity objects:
+ - Host Application Identity
+   - constraints
+     - namespace: `namespace-1`
+ - Authentication request Application Identity
+   - constraints
+     - namespace: `namespace-2`
+     
+All we need to do is iterate over the Authentication request Application Identity
+`constraints` hash and verify that all of its fields have the same
+ values as the corresponding `constraints` hash fields in the Host
+  Application Identity object.
+  
+In this case the general class will be defined as follows:
+- input
+  - Host Application Identity object
+  - Authentication request Application Identity object
+- output
+  - Nothing if validation succeeded, InvalidApplicationIdentity error if it
+   failed
+   
+### Implementation
+
+In conclusion, our general ValidateApplication class will need the following
+inputs:
+  - Authenticator name (e.g 'authn-azure')
+  - Authenticator's service-id (e.g 'prod')
+  - host's annotations
+  - Different lists of annotation names
+    - permitted
+    - required
+    - non-permitted combinations
+  - Authentication request Application Identity object
+  
+It will perform the following steps:
+  1. Build the Host Application Identity object using the:
```

Comment:
Here do you mean a completely, separate object that will have the attributes listed below?
In other words an object that has an authenticator name, service-id, host-annotations attributes will be dynamically filled based on the incoming request and the list of permitted annotations predefined?
So before runtime, the Azure Authentication Identity object would look like:
```
    authenticator name: null
    authenticator service-id: null
    host's annotations: null
    list of permitted annotation: subscription id, resource group, etc.
```
and at runtime:
```
    authenticator name: authn-azure
    authenticator service-id: oren-azure
    host's annotations: subscription-id: my-great-subscription-id, resource-group: my-resource-group
    list of permitted annotation: subscription id, resource group, etc.
```

Reply:
No. this class is described above like this:
```
- input
  - authenticator name
  - authenticator service-id
  - annotations
  - List of relevant annotations (e.g `[subscription-id, resource-group
  , system-assigned-identity, user-assigned-identity]
  ` for `authn
  -azure`)
- output
  - `constraints` hash.
```

It will have only one attribute - `constraints`